### PR TITLE
migrate the mac offical binary build to VSTS mac hosted preview

### PR DESCRIPTION
**Pull request by TravisEz13**
## PR Summary

migrate the mac offical binary build to VSTS mac hosted preview
    - create a PowerShellPackage...ps1 that works with these VMs (vs our custom VM)
    - create a script to create the \powershell folder with correct permissions
    - a yml VSTS build definition

#6355 should be merged first

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+if($Build.IsPresent)
+{
+    # cleanup the folder but don't delete it or the build agent will loose ownership of the folder
+    Write-Verbose -Message "cleaning /PowerShell" -Verbose
+    Get-ChildItem -Path /PowerShell/* -Attributes Hidden,Normal,Directory | Remove-Item -Recurse -Force
+
+    dir env:
+    # clone the repositor to the location we must build from
+    Write-Verbose -Message "cloaning to /PowerShell" -Verbose
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: > cloaning 

Typo

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    $repoRoot = $location
+}
+
+if($Build.IsPresent)
+{
+    # cleanup the folder but don't delete it or the build agent will loose ownership of the folder
+    Write-Verbose -Message "cleaning /PowerShell" -Verbose
+    Get-ChildItem -Path /PowerShell/* -Attributes Hidden,Normal,Directory | Remove-Item -Recurse -Force
+
+    dir env:
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: I guess this is for diagnostics purpose?

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    # clone the repositor to the location we must build from
```
Comment (adityapatwardhan) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: typo: repositor -> repository

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 2 above)
Comment (adityapatwardhan) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: Do we need this or this is a debug statement.

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    Import-Module "$repoRoot/tools/packaging"
+    Sync-PSTags -AddRemoteIfMissing
+
+    if($BootStrap.IsPresent)
+    {
+        Start-PSBootstrap -Package
+        Write-Verbose -Message "Installing fpm..." -Verbose
+        Start-NativeExecution { sudo gem install fpm -v 1.8.1 }
+        Write-Verbose -Message "Installing ronn..." -Verbose
+        Start-NativeExecution { sudo gem install ronn }
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: Won't `Start-PSBootstrap` install `fpm` and `ronn`?

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+if($Build.IsPresent)
+{
+    # cleanup the folder but don't delete it or the build agent will loose ownership of the folder
+    Write-Verbose -Message "cleaning /PowerShell" -Verbose
+    Get-ChildItem -Path /PowerShell/* -Attributes Hidden,Normal,Directory | Remove-Item -Recurse -Force
+
+    dir env:
+    # clone the repositor to the location we must build from
+    Write-Verbose -Message "cloaning to /PowerShell" -Verbose
+    git clone $location /PowerShell
```
Comment (adityapatwardhan) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: should be do a recursive clone?

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    if($Build.IsPresent)
+    {
+        Start-PSBuild -Crossgen -PSModuleRestore @releaseTagParam
+
+        Start-PSPackage @releaseTagParam
+        switch ($ExtraPackage)
+        {
+            "zip" { Start-PSPackage -Type zip @releaseTagParam }
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: We don't build `zip` package for macOS.

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 1 above)
Comment (adityapatwardhan) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: typo: cloaning -> cloning

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    }
+}
+
+
+Push-Location
+try {
+    Write-Verbose -Message "Init..." -Verbose
+    Set-Location $repoRoot
+    git submodule update --init --recursive --quiet
+    Import-Module "$repoRoot/build.psm1"
```
Comment (adityapatwardhan) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: If the VM is reused, we should use `-Force`

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    Import-Module "$repoRoot/tools/packaging"
```
Comment (adityapatwardhan) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: If the VM is reused, we should use `-Force`

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    }
+}
+finally
+{
+    Pop-Location
+}
+
+if($Build.IsPresent)
+{
+    $linuxPackages = Get-ChildItem "$repoRoot/powershell*" -Include *.pkg,*.zip,*.tar.gz
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: Maybe use `$macPackages`? And maybe `*.zip` can be removed.

File: tools/releaseBuild/macOS/vsts.yml [hunk 10]
```diff
@@ -0,0 +1,26 @@
+steps:
+- task: ShellScript@2
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/vsts.yml]: Curious, why all tasks here have the same name?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: tools/releaseBuild/macOS/vsts.yml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  displayName: 'Create /PowerShell'
+- task: ShellScript@2
+  inputs:
+    scriptPath: 'tools/releaseBuild/macOS/PowerShellPackageVsts.sh'
+    args: '-location $(Build.SourcesDirectory) -BootStrap'
+  displayName: 'Bootstrap VM'
+- task: ShellScript@2
+  inputs:
+    scriptPath: 'tools/releaseBuild/macOS/PowerShellPackageVsts.sh'
+    args: '-ReleaseTag $(ReleaseTagVar) -Destination $(System.ArtifactsDirectory) -ExtraPackage "zip","tar" -location $(Build.SourcesDirectory) -Build'
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/vsts.yml]: Do we need to `zip` here?

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 2 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: will remove

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 5 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: don't need to for official builds

Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: plus it is never quiet.

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 7 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: VM is reused but not the process.

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 8 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: VM is reused but not the process.

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 2 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: I will remove

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 4 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: No, it doesn't run it with sudo and sudo is required

File: tools/releaseBuild/macOS/vsts.yml (hunk 10 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/vsts.yml]: This is the name of the `<type>@<version>` of the task .   We are running the same task with different parameters.  So the name is the same.

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 3 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: fixed

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 1 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: fixed

File: tools/releaseBuild/macOS/vsts.yml (hunk 11 above)
Comment (TravisEz13) [reply tools/releaseBuild/macOS/vsts.yml]: fixed

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 (hunk 4 above)
Comment (daxian-dbw) [reply tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: Add a comment as we discussed.

File: tools/releaseBuild/macOS/PowerShellPackageVsts.ps1 [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    }
+}
+finally
+{
+    Pop-Location
+}
+
+if($Build.IsPresent)
+{
+    $linuxPackages = Get-ChildItem "$repoRoot/powershell*" -Include *.pkg,*.tar.gz
```
Comment (daxian-dbw) [inline tools/releaseBuild/macOS/PowerShellPackageVsts.ps1]: The variable name might be changed to `$macPackages`.

File: tools/releaseBuild/macOS/vsts.yml (hunk 10 above)
Comment (daxian-dbw) [reply tools/releaseBuild/macOS/vsts.yml]: A comment would be great.

**Review by daxian-dbw — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Comment by TravisEz13** [comment]
migrate the mac official binary build to VSTS mac hosted preview
- create a PowerShellPackage...ps1 that works with these VMs (vs our custom VM)
- create a script to create the \powershell folder with correct permissions
- a YML VSTS build definition

**Pull request merged**